trx_file_memmap/cpp/src/trx.cpp

```diff
@@ -0,0 +1,361 @@
+
+		// TODO: won't work on windows and not validating OS type
+		std::string base = get_base("/", filename);
+
+		size_t num_splits = std::count(base.begin(), base.end(), '.');
+		int dim;
+
+		if (num_splits != 1 and num_splits != 2)
+		{
+			throw std::invalid_argument("Invalid filename");
+		}
+
+		std::string ext = get_ext(filename);
+
+		base = base.substr(0, base.length() - ext.length());
+
+		if (num_splits == 1)
+		{
+			dim = 1;
+		}
+		else
+		{
+			int pos = base.find_last_of(".");
+			dim = std::stoi(base.substr(pos + 1, base.size()));
+			base = base.substr(0, pos);
+		}
+
+		bool is_valid = _is_dtype_valid(ext);
+
+		if (is_valid == false)
+		{
+			// TODO: make formatted string and include provided extension name
+			throw std::invalid_argument("Unsupported file extension");
+		}
+
+		std::tuple<std::string, int, std::string> output{base, dim, ext};
+
+		return output;
+	}
+
+	bool _is_dtype_valid(std::string &ext)
+	{
+		if (ext.compare("bit") == 0)
+			return true;
+		if (std::find(trxmmap::dtypes.begin(), trxmmap::dtypes.end(), ext.substr(1)) != trxmmap::dtypes.end())
+			return true;
+		return false;
+	}
+
+	json load_header(zip_t *zfolder)
+	{
+		// load file
+		zip_file_t *zh = zip_fopen(zfolder, "header.json", ZIP_FL_UNCHANGED);
+
+		// read data from file in chunks of 255 characters until data is fully loaded
+		int buff_len = 255 * sizeof(char);
+		char *buffer = (char *)malloc(buff_len);
+
+		std::string jstream = "";
+		zip_int64_t nbytes;
+		while ((nbytes = zip_fread(zh, buffer, buff_len - 1)) > 0)
+		{
+			if (buffer != NULL)
+			{
+				jstream += string(buffer, nbytes);
+			}
+		}
+
+		free(zh);
+		free(buffer);
+
+		// convert jstream data into Json.
+		auto root = json::parse(jstream);
+		return root;
+	}
+
+	int load_from_zip(const char *path)
+	{
+		int *errorp;
+		zip_t *zf = zip_open(path, 0, errorp);
+		json header = load_header(zf);
+
+		std::map<std::string, std::tuple<int, int>> file_pointer_size;
+		int global_pos = 0;
+		int mem_address = 0;
+
+		int num_entries = zip_get_num_entries(zf, ZIP_FL_UNCHANGED);
+
+		for (int i = 0; i < num_entries; ++i)
+		{
+			std::string elem_filename = zip_get_name(zf, i, ZIP_FL_UNCHANGED);
+
+			size_t lastdot = elem_filename.find_last_of(".");
+
+			if (lastdot == std::string::npos)
+				continue;
+			std::string ext = elem_filename.substr(lastdot + 1, std::string::npos);
+
+			if (ext.compare("bit") == 0)
+				ext = "bool";
+
+			// get file stats
+			zip_stat_t sb;
+
+			if (zip_stat(zf, elem_filename.c_str(), ZIP_FL_UNCHANGED, &sb) != 0)
+			{
+				return 1;
+			}
+
+			ifstream file(path, ios::binary);
+			file.seekg(global_pos);
+			mem_address = global_pos;
+
+			unsigned char signature[4] = {0};
+			const unsigned char local_sig[4] = {0x50, 0x4b, 0x03, 0x04};
+			file.read((char *)signature, sizeof(signature));
+
+			if (memcmp(signature, local_sig, sizeof(signature)) == 0)
+			{
+				global_pos += 30;
+				global_pos += sb.comp_size + elem_filename.size();
+			}
```

Comment:
I may not have completely understood this part- so if the file within the .trx is a zip file, it should skip the zip header? wouldn't it be relevant information for libzip or wouldn't libzip/trxlib assume that the file cursor starts at 0?